terraform/transform_reference.go

```diff
@@ -391,10 +369,44 @@ func (m ReferenceMap) dependsOn(g *Graph, depender graphNodeDependsOn) ([]dag.Ve
 	return res, fromModule || fromParentModule
 }
 
+// Return extra depends_on references if this is a data source.
+// For data sources we implicitly treat references to managed resources as
+// depends_on entries. If a data source references a managed resource, even if
+// that reference is resolvable, it stands to reason that the user intends for
+// the data source to require that resource in some way.
+func (m ReferenceMap) dataDependsOn(depender graphNodeDependsOn) []*addrs.Reference {
+	var refs []*addrs.Reference
+	if n, ok := depender.(GraphNodeConfigResource); ok &&
+		n.ResourceAddr().Resource.Mode == addrs.DataResourceMode {
+		for _, r := range depender.References() {
+
+			var resAddr addrs.Resource
+			switch s := r.Subject.(type) {
+			case addrs.Resource:
+				resAddr = s
+			case addrs.ResourceInstance:
+				resAddr = s.Resource
+				r.Subject = resAddr
```

Comment:
More a question: reading this movement from the previous block, this appears to be the addition that has changed things (changing the subject). What happened here? Reading the PR summary, I was looking for some change to reading `ManagedResourceMode` but that behavior hasn't changed (but please correct my reading!): you go from checking `==` for `ManagedResourceMode` to `!=` and a `continue` ... so this behavior is the same as the earlier block.

Reply:
I think this is the relevant bit in the PR summary?  

> we need to
re-target that reference to the containing resource for planning

But I'm also interested in this change from == to != ... is it doing anything I'm not seeing? (not saying to revert this part of the change, this `continue` might make things make more logical sense to future readers)

Reply:
Yes, the `r.Subject = resAddr` is key change here. The problem was that when the reference was to an indexed resource instance, that wouldn't match up with the resource addresses used for depends_on when building the plan graph, since there are no individual instances at that point. This prevented the data source from being connected to the resource during planning, and allowing the data source to be read earlier than expected.

You correct that the `resAddr.Mode != addrs.ManagedResourceMode` logic didn't change. I believe I had some more code in here that made a difference during the refactor (and made it longer, instigating moving it to another method), but the append logic did end up being the same in the end.